File: video_engine/lib/steps/s10_inpaint.dart
import 'dart:io';
import 'package:path/path.dart' as p;
import '../core/job_state.dart';
import '../core/step_base.dart';
import '../utils/ffmpeg_utils.dart';

class StepInpaint extends StepBase {
  @override
  String get stepId => 's10_inpaint';

  @override
  List<String> get dependsOn => const ['s02_demux', 's03_subtitle_detect'];

  @override
  List<String> get stepConfigKeys => const [
        'inpaint_engine',
        'inpaint_method',
        'inpaint_region',
        'inpaint_blur_radius',
        'watermark_enabled',
        'watermark_region',
        'watermark_image',
        'watermark_text',
        'watermark_font_name',
        'watermark_font_color',
        'watermark_opacity',
        'watermark_blur_bg',
      ];

  @override
  Future<Map<String, dynamic>> run(
    Directory workspace,
    Map<String, dynamic> config,
    JobState jobState,
  ) async {
    final demuxInfo = jobState.getStepOutput('s02_demux') ?? {};
    final detectInfo = jobState.getStepOutput('s03_subtitle_detect') ?? {};
    final probeInfo = jobState.getStepOutput('s01_probe') ?? {};

    final videoStream = demuxInfo['video_stream'] as String?;
    if (videoStream == null || !File(videoStream).existsSync()) {
      throw ArgumentError("Video stream '$videoStream' not found.");
    }

    final cleanVideo = File(p.join(workspace.path, 'clean_video.mp4'));
    final inpaintCfg = config['inpaint'] is Map ? (config['inpaint'] as Map) : {};
    final inpaintEngine = (config['inpaint_engine'] ?? inpaintCfg['engine'] ?? 'apple_vision_inpaint').toString().toLowerCase();
    final inpaintMethod = (config['inpaint_method'] ?? inpaintCfg['method'] ?? 'vertical_gradient').toString().toLowerCase();

    // 1. If inpaint engine is box_color: SubBox is drawn via ASS Layer 0 in s09/s11
    if (inpaintEngine == 'box_color' || inpaintEngine == 'box') {
      if (videoStream != cleanVideo.path) {
        File(videoStream).copySync(cleanVideo.path);
      }
      return {
        'skipped': true,
        'clean_video': cleanVideo.path,
      };
    }

    final videoWidth = (probeInfo['width'] as num?)?.toInt() ?? 1080;
    final videoHeight = (probeInfo['height'] as num?)?.toInt() ?? 1920;

    List<double> region = [0.75, 0.05, 0.95, 0.95];
    final cfgRegion = config['inpaint_region'];
    if (cfgRegion is List && cfgRegion.length == 4) {
      region = cfgRegion.map((e) => double.tryParse(e.toString()) ?? 0.0).toList();
    } else if (detectInfo['burnin_region'] is List && (detectInfo['burnin_region'] as List).length == 4) {
      region = (detectInfo['burnin_region'] as List).map((e) => double.tryParse(e.toString()) ?? 0.0).toList();
    }

    final paddingY = double.tryParse((config['inpaint_padding_y'] ?? inpaintCfg['padding_y'] ?? config['blur_box_padding_y'])?.toString() ?? '0.005') ?? 0.005;

    // Apply padding_y
    final ymin = (region[0] - paddingY).clamp(0.0, 1.0);
    final xmin = region[1].clamp(0.0, 1.0);
    final ymax = (region[2] + paddingY).clamp(0.0, 1.0);
    final xmax = region[3].clamp(0.0, 1.0);
    final paddedRegion = [ymin, xmin, ymax, xmax];

    final tempInpainted = File(p.join(workspace.path, 'temp_inpainted.mp4'));
    final inpaintOutPath = tempInpainted.path;
    bool inpaintDone = false;

    // 2. Native Directional Vertical Gradient Inpaint (100% parity with Python apple_vision_inpaint)
    if (inpaintEngine == 'apple_vision_inpaint' || inpaintEngine == 'apple_vision' || inpaintMethod == 'vertical_gradient') {
      final inpaintBin = _resolveInpaintBinary();
      if (inpaintBin != null && File(inpaintBin).existsSync()) {
        final fps = (probeInfo['fps'] as num?)?.toDouble() ?? 30.0;
        final bitrate = config['video_bitrate']?.toString() ?? '4.0M';

        final args = [
          videoStream,
          inpaintOutPath,
          paddedRegion[0].toString(),
          paddedRegion[1].toString(),
          paddedRegion[2].toString(),
          paddedRegion[3].toString(),
          videoWidth.toString(),
          videoHeight.toString(),
          fps.toString(),
          bitrate,
        ];

        final res = await Process.run(inpaintBin, args);
        if (res.exitCode == 0 && tempInpainted.existsSync() && tempInpainted.lengthSync() > 1000) {
          inpaintDone = true;
        }
      }
    }

    // 3. Fallback / ffmpeg_blur Mode
    if (!inpaintDone) {
      final topPx = (paddedRegion[0] * videoHeight).round();
      final leftPx = (paddedRegion[1] * videoWidth).round();
      final bottomPx = (paddedRegion[2] * videoHeight).round();
      final rightPx = (paddedRegion[3] * videoWidth).round();
      final boxWidth = rightPx - leftPx;
      final boxHeight = bottomPx - topPx;

      final blurRadius = (config['inpaint_blur_radius'] as num?)?.toInt() ?? 15;
      final ffmpegBin = FFmpegUtils.resolveBinary('ffmpeg');
      final filter = '[0:v]crop=$boxWidth:$boxHeight:$leftPx:$topPx,boxblur=$blurRadius[blur];[0:v][blur]overlay=$leftPx:$topPx[vfinal]';

      final args = [
        '-y',
        '-i', videoStream,
        '-filter_complex', filter,
        '-map', '[vfinal]',
        '-c:v', 'h264_videotoolbox',
        '-b:v', config['video_bitrate']?.toString() ?? '4.0M',
        '-pix_fmt', 'yuv420p',
        '-an',
        inpaintOutPath,
      ];

      final res = await Process.run(ffmpegBin, args);
      if (res.exitCode == 0 && tempInpainted.existsSync() && tempInpainted.lengthSync() > 1000) {
        inpaintDone = true;
      }
    }

    final videoForWatermark = inpaintDone ? inpaintOutPath : videoStream;

    // 4. Apply watermark if enabled
    final wmApplied = await FFmpegUtils.applyWatermark(
      inputVideo: videoForWatermark,
      outputVideo: cleanVideo.path,
      config: config,
      width: videoWidth,
      height: videoHeight,
    );

    if (!wmApplied) {
      if (videoForWatermark != cleanVideo.path) {
        File(videoForWatermark).copySync(cleanVideo.path);
      }
    }

    if (tempInpainted.existsSync()) {
      try { tempInpainted.deleteSync(); } catch (_) {}
    }

    return {
      'clean_video': cleanVideo.path,
      'watermark_applied': wmApplied,
    };
  }

  static String? _resolveInpaintBinary() {
    final home = Platform.environment['HOME'] ?? '';
    final candidates = [
      p.join(home, 'Library', 'Application Support', 'SubVideo', 'engine', 'sub_video_inpaint'),
      p.join(File(Platform.resolvedExecutable).parent.path, 'sub_video_inpaint'),
      '/Users/voquyt/Documents/projects/video/Sub-Video/rust_native/target/release/sub_video_inpaint',
      p.join(home, 'Documents', 'projects', 'video', 'Sub-Video', 'rust_native', 'target', 'release', 'sub_video_inpaint'),
    ];

    for (final c in candidates) {
      if (File(c).existsSync()) {
        try {
          Process.runSync('chmod', ['+x', c]);
        } catch (_) {}
        return c;
      }
    }
    return null;
  }
}

File: py_engine/steps/s10_inpaint.py
import json
import shutil
import sys
from pathlib import Path
from typing import Any, Dict

from core.plugin_loader import PluginLoader
from core.step_base import StepBase
from utils.ffmpeg_utils import FFmpegUtils


class StepInpaint(StepBase):
    step_id = "s10_inpaint"
    depends_on = ["s02_demux", "s03_subtitle_detect"]
    STEP_CONFIG_KEYS = [
        "show_subtitle", "inpaint", "inpaint_show_box", "inpaint_method", "inpaint_region", "inpaint_color", "blur_radius", 
        "video_bitrate", "blur_box_padding_y", "watermark_enable", "watermark_region", 
        "watermark_image", "watermark_text", "watermark_font_name", "watermark_blur_bg", 
        "watermark_opacity", "watermark_font_color"
    ]

    def run(self, workspace: Path, config: Dict[str, Any], job_state: Any) -> Dict[str, Any]:
        detect_info = job_state.get_step_output("s03_subtitle_detect") or {}
        ocr_info = job_state.get_step_output("s06_ocr") or {}
        demux_info = job_state.get_step_output("s02_demux") or {}
        probe_info = job_state.get_step_output("s01_probe") or {}

        mode = detect_info.get("mode")
        v_cand = demux_info.get("video_stream")
        if v_cand and Path(v_cand).exists():
            input_video = Path(v_cand)
        elif (workspace / "demux" / "video_stream.mp4").exists():
            input_video = workspace / "demux" / "video_stream.mp4"
        elif (workspace / "video_stream.mp4").exists():
            input_video = workspace / "video_stream.mp4"
        elif Path(job_state.data.get("input_video", "")).exists():
            input_video = Path(job_state.data["input_video"])
        else:
            input_video = Path(v_cand) if v_cand else (workspace / "demux" / "video_stream.mp4")

        clean_video = workspace / "clean_video.mp4"
        temp_inpainted = workspace / "temp_inpainted.mp4"

        inpaint_cfg = config.get("inpaint") if isinstance(config.get("inpaint"), dict) else {}
        ocr_only = config.get("ocr_only", False)
        show_sub = config.get("show_subtitle", True)
        show_box = bool(config.get("inpaint_show_box") if config.get("inpaint_show_box") is not None else inpaint_cfg.get("show_box", True))
        override_region = config.get("inpaint_region") or inpaint_cfg.get("region")
        need_inpaint = show_box

        if need_inpaint:
            # Calculate inpaint region: manual config region > auto-detect burnin > ocr detected region > fallback
            raw_region = override_region or detect_info.get("burnin_region") or ocr_info.get("detected_sub_region")
            padding_y = float(inpaint_cfg.get("padding_y") or config.get("blur_box_padding_y") or 0.01)

            if raw_region and len(raw_region) == 4:
                if not override_region:
                    ymin, xmin, ymax, xmax = raw_region
                    padded_ymin = max(0.0, ymin - padding_y)
                    padded_ymax = min(1.0, ymax + padding_y)
                    region = [round(padded_ymin, 3), xmin, round(padded_ymax, 3), xmax]
                else:
                    region = raw_region
            else:
                region = None

            segments = None
            for cand_name in ["s08c_timing.json", "s08_translation.json", "s07_transcript.json"]:
                cand_file = workspace / cand_name
                if cand_file.exists():
                    try:
                        with open(cand_file, "r", encoding="utf-8") as f:
                            segments = json.load(f)
                            if segments:
                                break
                    except Exception:
                        pass

            if not region and segments:
                # Auto-calculate bounding box enclosing all OCR detected subtitle texts with 10% side margins (0.1 -> 0.9)
                bboxes = [s["bbox"] for s in segments if "bbox" in s and isinstance(s["bbox"], list) and len(s["bbox"]) == 4]
                if bboxes:
                    auto_ymin = max(0.0, min(b[0] for b in bboxes) - padding_y)
                    auto_ymax = min(1.0, max(b[2] for b in bboxes) + padding_y)
                    region = [round(auto_ymin, 3), 0.1, round(auto_ymax, 3), 0.9]

            if not region:
                region = [0.80, 0.05, 0.95, 0.95]

            inpaint_engine = str(inpaint_cfg.get("engine") or config.get("inpaint") or "ffmpeg_blur").replace("-", "_").lower()
            inpaint_color = str(inpaint_cfg.get("color") or config.get("inpaint_color") or "transparent").strip().lower()
            is_solid_box = (inpaint_engine in ["box_color", "box"]) or (inpaint_color not in ["transparent", "", "none"])

            # When a solid colored box is requested (e.g. black, white), ffmpeg_blur's drawbox is optimal (~0.5s)
            if is_solid_box:
                inpaint_plugin_name = "ffmpeg_blur"
            elif inpaint_engine == "opencv":
                inpaint_plugin_name = "opencv_inpaint"
            elif inpaint_engine in ("apple_vision", "apple_vision_inpaint", "applevision"):
                inpaint_plugin_name = "ffmpeg_blur" if sys.platform == "win32" else "apple_vision_inpaint"
            elif inpaint_engine in ("blur", "ffmpeg_blur", "boxblur"):
                inpaint_plugin_name = "ffmpeg_blur"
            else:
                inpaint_plugin_name = "ffmpeg_blur"

            # Merge and preserve all engine-specific configurations to prevent miss-config
            merged_config = dict(config)
            merged_config["inpaint_region"] = region
            merged_config["inpaint_method"] = inpaint_cfg.get("method") or config.get("inpaint_method") or "vertical_gradient"
            merged_config["blur_radius"] = inpaint_cfg.get("blur_radius") or config.get("blur_radius") or 15
            merged_config["blur_box_padding_y"] = padding_y
            if "box" in inpaint_cfg and isinstance(inpaint_cfg["box"], dict):
                merged_config["inpaint_box"] = inpaint_cfg["box"]

            inpaint_plugin = PluginLoader.load_plugin("inpaint", inpaint_plugin_name, merged_config)
            inpaint_segments = segments

            # Validate watermark config strictly before processing
            wm_cfg = FFmpegUtils.validate_watermark_config(config, workspace=workspace)

            # 1-Pass Video Filtergraph: ffmpeg_blur merges inpaint boxblur + watermark into single encode
            try:
                inpaint_plugin.remove_subtitles(
                    input_video,
                    region,
                    clean_video,
                    segments=inpaint_segments,
                    watermark_config=wm_cfg
                )
                inpaint_applied = True
                wm_applied = bool(wm_cfg and wm_cfg.get("enabled"))
            except TypeError:
                # Fallback for third-party inpaint plugins without watermark_config support
                try:
                    inpaint_plugin.remove_subtitles(input_video, region, temp_inpainted, segments=inpaint_segments)
                except TypeError:
                    inpaint_plugin.remove_subtitles(input_video, region, temp_inpainted)
                inpaint_applied = True

                if wm_cfg:
                    video_width = int(probe_info.get("width") or 1920)
                    video_height = int(probe_info.get("height") or 1080)
                    config_ctx = config.copy() if hasattr(config, "copy") else dict(config)
                    config_ctx["workspace_dir"] = str(workspace)
                    wm_applied = FFmpegUtils.apply_watermark(
                        temp_inpainted,
                        clean_video,
                        config_ctx,
                        width=video_width,
                        height=video_height
                    )
                else:
                    shutil.copy(str(temp_inpainted), str(clean_video))
                    wm_applied = False
        else:
            inpaint_applied = False
            # Check if watermark is permitted and requested even when inpaint is skipped
            wm_cfg = FFmpegUtils.validate_watermark_config(config, workspace=workspace)
            if wm_cfg:
                video_width = int(probe_info.get("width") or 1920)
                video_height = int(probe_info.get("height") or 1080)
                config_ctx = config.copy() if hasattr(config, "copy") else dict(config)
                config_ctx["workspace_dir"] = str(workspace)
                wm_applied = FFmpegUtils.apply_watermark(
                    input_video,
                    clean_video,
                    config_ctx,
                    width=video_width,
                    height=video_height
                )
            else:
                shutil.copy(str(input_video), str(clean_video))
                wm_applied = False

        if temp_inpainted.exists():
            temp_inpainted.unlink(missing_ok=True)

        return {
            "inpaint_applied": inpaint_applied,
            "watermark_applied": wm_applied,
            "clean_video": str(clean_video)
        }
